### PR TITLE
more convenient constructor for AASequence in Python

### LINKED ISSUE #7647: Wrap static AASequence::fromString() for easier use in python
e.g.., to just call
a = AASequence(b"PEPTIDER");
a.toMZ(2);

### PR DESCRIPTION
**Pull request by timosachsenberg**
## Checklist
- [ ] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [ ] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

___

### **PR Type**
enhancement, tests

___

### **Description**
- Introduced new constructors for the `AASequence` class to allow initialization from `String` and `char*`.
- Updated the `AASequence` header to include explicit constructors and made the move assignment operator `noexcept`.
- Added equality operators to the `AASequence` Cython wrapper.
- Implemented unit tests to verify the functionality of the new constructors and operators.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>AASequence.cpp</strong><dd><code>Add new constructors for AASequence class</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/openms/source/CHEMISTRY/AASequence.cpp

<li>Added new constructors for <code>AASequence</code> from <code>String</code> and <code>char*</code>.<br> <li> Initialized <code>n_term_mod_</code> and <code>c_term_mod_</code> to <code>nullptr</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7650/files#diff-0040f2beba30f4f40b8bdbf21c98b822e01a3b334a13e7a5e80b72d7d99bf3e7">+15/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>AASequence.pxd</strong><dd><code>Extend AASequence with new constructor and operators</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/pyOpenMS/pxds/AASequence.pxd

<li>Added new constructor for <code>AASequence</code> from <code>String</code>.<br> <li> Added equality operators for <code>AASequence</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7650/files#diff-1f4229a559cf1ba0460b73d7c7cd52d3b56f7bf230d1237aac897eda16b7c3af">+6/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>AASequence.h</strong><dd><code>Update AASequence interface with new constructors</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/openms/include/OpenMS/CHEMISTRY/AASequence.h

<li>Added explicit constructors for <code>AASequence</code> from <code>String</code> and <code>char*</code>.<br> <li> Made move assignment operator <code>noexcept</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7650/files#diff-df09f7e70ffbe48753a3299ef6fea8154ee65547bb05d52852b7d467a6798f88">+9/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test000.py</strong><dd><code>Add unit tests for new AASequence constructors</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/pyOpenMS/tests/unittests/test000.py

<li>Added unit tests for new <code>AASequence</code> constructors.<br> <li> Verified equality and string conversion methods.<br>

</details>

  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7650/files#diff-9b64f80ffadfaff913d4cd4a67264c1b950fc47ad83ebf98bcaf1431d6479354">+10/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

**Review by jcharkow — APPROVED** [review]
Looks good! This is a much more pythonic way of initiating the object.

One thing, is it possible to throw an error on the python side if a user does 
AASequence().from_string("PEPTIDER"). E.g. is there a way to ensure that a static method is not called on an actual object from the python side? 

As a side note would it be possible to do something similar to this for the MSExperiment object? E.g. Instead of 
```
exp = oms.MSExperiment()
oms.MzMLFile().load("test.mzML", exp)
```

```
exp = oms.MSExperiment("test.mzML")
```

Also once this is merged we should update the documentation to reflect this new way as I think it will be more straightforward for new python users.

**Comment by jcharkow** [comment]
Also I am not sure if the codium AI suggestions are relevant or not

**Comment by timosachsenberg** [comment]
> Looks good! This is a much more pythonic way of initiating the object.
> 
> One thing, is it possible to throw an error on the python side if a user does AASequence().from_string("PEPTIDER"). E.g. is there a way to ensure that a static method is not called on an actual object from the python side?

I don't think so. But maybe we just don't wrap the static function then?

> As a side note would it be possible to do something similar to this for the MSExperiment object? E.g. Instead of
> 
> ```
> exp = oms.MSExperiment()
> oms.MzMLFile().load("test.mzML", exp)
> ```
> 
> ```
> exp = oms.MSExperiment("test.mzML")
> ```
> 
> Also once this is merged we should update the documentation to reflect this new way as I think it will be more straightforward for new python users.

From a C++ perspective, adding serialization logic to the MSExperiment class would be a poor design choice. The reason is that this approach implies the data model must "know" the details of its serialization. Ideally, the dependency should work in the opposite direction: the serialization should understand the data model, not the other way around.
In languages that allow "monkey patching" / "dynamic method addition" may still do it the other way around, simulating a kind of "plugin" mechanism to add behavior at runtime. Doesn't work well with our design though.
BUT: we could in principle add it to https://github.com/OpenMS/OpenMS/blob/develop/src/pyOpenMS/pyopenms/_dataframes.py where we add functionality to an existing class. 
Maybe worth discussing?

File: src/openms/include/OpenMS/CHEMISTRY/AASequence.h [hunk 1]
```diff
@@ -597,13 +597,19 @@ namespace OpenMS
     static AASequence fromString(const char* s,
                                  bool permissive = true);
 
+    /// constructor from String
+    explicit AASequence(const String& s);
+
+    /// constructor from C string
+    explicit AASequence(const char* s);
+
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/CHEMISTRY/AASequence.h]:
```suggestion
    /// @brief constructor from String
    /// @param s A String representing the amino acid sequence
    explicit AASequence(const String& s);

    /// @brief constructor from C string
    /// @param s A C-style string representing the amino acid sequence
    explicit AASequence(const char* s);
```

**Comment by jcharkow** [comment]
> > Looks good! This is a much more pythonic way of initiating the object.
> > One thing, is it possible to throw an error on the python side if a user does AASequence().from_string("PEPTIDER"). E.g. is there a way to ensure that a static method is not called on an actual object from the python side?
> 
> I don't think so. But maybe we just don't wrap the static function then?
> 
> > As a side note would it be possible to do something similar to this for the MSExperiment object? E.g. Instead of
> > ```
> > exp = oms.MSExperiment()
> > oms.MzMLFile().load("test.mzML", exp)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> > exp = oms.MSExperiment("test.mzML")
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Also once this is merged we should update the documentation to reflect this new way as I think it will be more straightforward for new python users.
> 
> From a C++ perspective, adding serialization logic to the MSExperiment class would be a poor design choice. The reason is that this approach implies the data model must "know" the details of its serialization. Ideally, the dependency should work in the opposite direction: the serialization should understand the data model, not the other way around. In languages that allow "monkey patching" / "dynamic method addition" may still do it the other way around, simulating a kind of "plugin" mechanism to add behavior at runtime. Doesn't work well with our design though. BUT: we could in principle add it to https://github.com/OpenMS/OpenMS/blob/develop/src/pyOpenMS/pyopenms/_dataframes.py where we add functionality to an existing class. Maybe worth discussing?

I am in favour of removing the static function to avoid this error however this will cause breaking changes in existing pipelines so it will have to be well documented. E.g. This functionality is used in easypqp https://github.com/grosenberger/easypqp/blob/bd78bcc1837b1841c021dd6c8dcb1c579f19ef09/easypqp/convert.py#L820
Possibly we can add a deprecation warning and then deprecate it? I am not sure how easy this would be. 

I think that adding that functionality to the _dataframes.py class is worth discussing. I would be in favour of this provided it is well documented.  I am currently not a huge fan of how the _dataframes.py class is implemented because I think it is very confusing but I do not know if there is a better way to do it.

**Comment by jpfeuffer** [comment]
> Looks good! This is a much more pythonic way of initiating the object.
> 
> One thing, is it possible to throw an error on the python side if a user does AASequence().from_string("PEPTIDER"). E.g. is there a way to ensure that a static method is not called on an actual object from the python side?

You can do almost anything in Python. You could probably override the `__getattribute` function of classes with staticmethods and throw errors when users call them. But the question is why? You can do the same in C++ and the compiler does not necessarily stop you.
I think it is definitely not worth the effort (of either doing it for every class, or implementing something automatic in autowrap). There are probably a lot of nuances with extension types, inheritance etc.

**Comment by jpfeuffer** [comment]
> I think that adding that functionality to the _dataframes.py class is worth discussing. I would be in favour of this provided it is well documented. I am currently not a huge fan of how the _dataframes.py class is implemented because I think it is very confusing but I do not know if there is a better way to do it.

The problem is that you cannot monkey patch C extension types since their data (members, functions etc.) are immutable.
Therefore you have to inherit from them and add the additional functionality there. To be backwards-compatible, however, you need to rename the modified class to the original name and hope it behaves the same for all use cases (I am still not sure if it really does).
I also do not like it at all. Ideally extending our classes is done in the pyx but most people are not familiar with Cython, so it was not done.

**Comment by jcharkow** [comment]
> > Looks good! This is a much more pythonic way of initiating the object.
> > One thing, is it possible to throw an error on the python side if a user does AASequence().from_string("PEPTIDER"). E.g. is there a way to ensure that a static method is not called on an actual object from the python side?
> 
> You can do almost anything in Python. You could probably override the `__getattribute` function of classes with staticmethods and throw errors when users call them. But the question is why? You can do the same in C++ and the compiler does not necessarily stop you. I think it is definitely not worth the effort (of either doing it for every class, or implementing something automatic in autowrap). There are probably a lot of nuances with extension types, inheritance etc.


I just do not think it is right for ```aa = AASequence().from_string("PEPTIDER")``` to exit with a non 0 exit status when in fact the sequence is not being set. It may lead to frustration with users.

I think this is one example that highlights that python is a second class citizen compared to c++ and throwing an error message when the output is not as expected is a step in the right direction towards reaching out to python developers. 

Maybe we do not need something for every class but for highly used classes, or just as they are noticed in pyopenms it would be nice. The other thing that Timo mentioned is that I am not sure how prevalent this is because it would only occur with static methods.

**Comment by jcharkow** [comment]
> > > Looks good! This is a much more pythonic way of initiating the object.
> > > One thing, is it possible to throw an error on the python side if a user does AASequence().from_string("PEPTIDER"). E.g. is there a way to ensure that a static method is not called on an actual object from the python side?
> > 
> > 
> > You can do almost anything in Python. You could probably override the `__getattribute` function of classes with staticmethods and throw errors when users call them. But the question is why? You can do the same in C++ and the compiler does not necessarily stop you. I think it is definitely not worth the effort (of either doing it for every class, or implementing something automatic in autowrap). There are probably a lot of nuances with extension types, inheritance etc.
> 
> I just do not think it is right for `aa = AASequence().from_string("PEPTIDER")` to exit with a non 0 exit status when in fact the sequence is not being set. It may lead to frustration with users.
> 
> I think this is one example that highlights that python is a second class citizen compared to c++ and throwing an error message when the output is not as expected is a step in the right direction towards reaching out to python developers.
> 
> Maybe we do not need something for every class but for highly used classes, or just as they are noticed in pyopenms it would be nice. The other thing that Timo mentioned is that I am not sure how prevalent this is because it would only occur with static methods.

Possibly the best option is to not wrap these functions in python if there is a possibility that they will not behave as expected.

**Comment by jpfeuffer** [comment]
> > > Looks good! This is a much more pythonic way of initiating the object.
> > > One thing, is it possible to throw an error on the python side if a user does AASequence().from_string("PEPTIDER"). E.g. is there a way to ensure that a static method is not called on an actual object from the python side?
> > 
> > You can do almost anything in Python. You could probably override the `__getattribute` function of classes with staticmethods and throw errors when users call them. But the question is why? You can do the same in C++ and the compiler does not necessarily stop you. I think it is definitely not worth the effort (of either doing it for every class, or implementing something automatic in autowrap). There are probably a lot of nuances with extension types, inheritance etc.
> 
> 
> I just do not think it is right for ```aa = AASequence().from_string("PEPTIDER")``` to exit with a non 0 exit status when in fact the sequence is not being set. It may lead to frustration with users.
> 
> I think this is one example that highlights that python is a second class citizen compared to c++ and throwing an error message when the output is not as expected is a step in the right direction towards reaching out to python developers. 
> 
> Maybe we do not need something for every class but for highly used classes, or just as they are noticed in pyopenms it would be nice. The other thing that Timo mentioned is that I am not sure how prevalent this is because it would only occur with static methods. 

How do you mean "the sequence is not set"? This should work, no? Since pyopenms is almost fully typed, you will see that it returns an AASequence object, so it makes sense that the initial instance (especially since it's a temporary one here) is not modified. But the returned object `aa` should be as one expects.

**Comment by jpfeuffer** [comment]
But in general, I am not against someone implementing safer python bindings,
as long as it's not toooo much overhead in maintenance and performance (which it should not be the case here). But I would personally not prioritize it over some other important things that might make people not even consider using OpenMS (other vendor file types, better ion mobility support, easier/more compatible installs)

**Comment by jcharkow** [comment]
Here is the full example (from the discord chat). I know I am not using pyopenms correctly in this example however without awareness to the underlying c++ implementation making this mistake is quite easy to do. But yes I agree creating safer pyopenms bindings does seem to be quite difficult and might not be as impactful as other things. It would be nice if there was an easy way to do this :) 

"""
Was using a part of pyopenms to compute m/z from a peptide sequence and a charge state. I do not use this a  lot  and thought I would share my experience. 

Basically given a peptide sequence and charge state I wanted to compute the m/z of the peptide sequence. 

With prior knowledge of how pyopenms works, my first attempt was the following:

import pyopenms as po
aaSeq = po.AASequence()
aaSeq.fromString(b'PEPTIDER')
aaSeq.getMZ(2) # returns 0.0


Luckily for this example, the documentation has an example of how to correctly do this, however I only know this because I know the documentation contains this example.

import pyopenms as po
aaSeq = po.AASequence.fromString(b'PEPTIDER')
aaSeq.getMZ(2) # returns 478.737817003521


From the c++ side it is clear this is a static method however for a python-focused developer or someone just getting started with pyopenms this can be confusing.

Finding all of these cases is not straightforward, subjective and would be a lot of work so I am not sure if it is worth the time it would take, just wanted to share my experience and see if anyone had any thoughts.
"""

**Comment by cbielow** [comment]
this unused return value problem is basically everywhere in Python.
See here https://discuss.python.org/t/proposal-typing-no-discard-a-decorator-to-indicate-that-the-return-value-should-not-be-discarded/33138
And it splits into three camps:
1) RTFM
2) give me a `no_discard` please to avoid it
3) that's only due to bad API design

I'm with @jpfeuffer here.

**Comment by poshul** [comment]
I've opened an issue to continue the discussion on the return value stuff #7651 . Outside of that do @jpfeuffer or @cbielow  have issues with the existing part of this patch?

File: src/openms/source/CHEMISTRY/AASequence.cpp [hunk 2]
```diff
@@ -1691,4 +1691,19 @@ namespace OpenMS
...
+  {
+    parseString_(s, *this, true);
+  }
+
+  AASequence::AASequence(const char* s) :
+    n_term_mod_(nullptr),
+    c_term_mod_(nullptr)
+  {
+    parseString_(s, *this, true);
+  }
```
Comment (cbielow) [inline src/openms/source/CHEMISTRY/AASequence.cpp]: can we put these at the top next to the default ctor?!
and ideally stay DRY by omitting
```
   n_term_mod_(nullptr),
    c_term_mod_(nullptr)
```
in all three cases (replace by ` ...n_term_mod_ = nullptr;` in header.

**Review by cbielow — APPROVED** [review]
ok, but a bit of cleaning up would be nice.


Also: what about not exposing the static methods in the Python interface or removing them entirely?

**Comment by timosachsenberg** [comment]
> ok, but a bit of cleaning up would be nice.
> 
> Also: what about not exposing the static methods in the Python interface or removing them entirely?

It is currently used by e.g. EasyPQP. I think we should deprecate it and remove it in a later version.

**Comment by jcharkow** [comment]
> > ok, but a bit of cleaning up would be nice.
> > Also: what about not exposing the static methods in the Python interface or removing them entirely?
> 
> It is currently used by e.g. EasyPQP. I think we should deprecate it and remove it in a later version.

I would be happy with this suggestion. I'm not sure if there is an easy way to mark deprecation, maybe just wrap this function manually to include a deprecation notice and then remove in a future version.

**Comment by timosachsenberg** [comment]
> > > ok, but a bit of cleaning up would be nice.
> > > Also: what about not exposing the static methods in the Python interface or removing them entirely?
> > 
> > 
> > It is currently used by e.g. EasyPQP. I think we should deprecate it and remove it in a later version.
> 
> I would be happy with this suggestion. I'm not sure if there is an easy way to mark deprecation, maybe just wrap this function manually to include a deprecation notice and then remove in a future version.

Yeah, I think this would also be an option.

**Comment by jcharkow** [comment]
@timosachsenberg will the deprecation note raise a warning? Otherwise it might go unnoticed

**Comment by timosachsenberg** [comment]
> @timosachsenberg will the deprecation note raise a warning? Otherwise it might go unnoticed

No. We could add it to the release notes. Not sure how to wrap it in the pyx file though.

File: src/pyOpenMS/pxds/AASequence.pxd [hunk 3]
```diff
@@ -136,7 +137,7 @@ cdef extern from "<OpenMS/CHEMISTRY/AASequence.h>" namespace "OpenMS":
 cdef extern from "<OpenMS/CHEMISTRY/AASequence.h>" namespace "OpenMS::AASequence":
         
         # static members
-        AASequence fromString(String s, bool permissive) except + nogil   # wrap-attach:AASequence wrap-as:fromStringPermissive
+        AASequence fromString(String s, bool permissive) except + nogil   # wrap-attach:AASequence wrap-as:fromStringPermissive wrap-doc:deprecated. Use AASequence(String) instead.
```
Comment (timosachsenberg) [inline src/pyOpenMS/pxds/AASequence.pxd:140]: I think we then also need to wrap the (String) version.  I think default parameter are not supported in pxd files.

**Comment by jcharkow** [comment]
> > @timosachsenberg will the deprecation note raise a warning? Otherwise it might go unnoticed
> 
> No. We could add it to the release notes. Not sure how to wrap it in the pyx file though.

Would be nice if that could be added to autowrap but for now this is fine

File: src/pyOpenMS/pxds/AASequence.pxd (hunk 3 above)
Comment (jcharkow) [reply src/pyOpenMS/pxds/AASequence.pxd:140]: I think I did that here? https://github.com/OpenMS/OpenMS/blob/688a53d1817fe23d65edfae15957efbbd3eee6e7/src/pyOpenMS/pxds/AASequence.pxd#L24

Comment (timosachsenberg) [reply src/pyOpenMS/pxds/AASequence.pxd:140]: Ah right  I did not see the previous change

**Comment by jpfeuffer** [comment]
Could you please add a complete deprecated directive such that it is properly parsed, recognized and prominently shown in our docs?
https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-deprecated

Real deprecation is coming with 3.13 as `@deprecated` decorator.  You could use typing_extension.deprecated right now but yes you either need to wrap it manually in pyx or add a new wrap-directive for autowrap (since cdef functions in pxds cannot be decorated arbitrarily.

Potentially the sphinx directive in the docstrings is enough for some IDEs to show a striked through function.

**Comment by timosachsenberg** [comment]
> Could you please add a complete deprecated directive such that it is properly parsed, recognized and prominently shown in our docs? https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-deprecated
> 
> Real deprecation is coming with 3.13 as `@deprecated` decorator. You could use typing_extension.deprecated right now but yes you either need to wrap it manually in pyx or add a new wrap-directive for autowrap (since cdef functions in pxds cannot be decorated arbitrarily.
> 
> Potentially the sphinx directive in the docstrings is enough for good IDEs

Does something multiline work?
```
            fromString(String seq) # wrap-attach:AASequence 
            # wrap-doc:
              #  :param seq: Construct AASequence from String (e.g., "PEPTIDER")
              #  .. deprecated:: 3.2
              #     Use :py:func:`AASequence(String)` instead.
              #  
```

**Comment by jpfeuffer** [comment]
You could actually do it right now.
When autowrap generates the pyi, you can check the docstring. If it contains `.. deprecated:` you add `@typing_extension.deprecated` before the type annotation.

**Comment by jpfeuffer** [comment]
> > Could you please add a complete deprecated directive such that it is properly parsed, recognized and prominently shown in our docs? https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-deprecated
> > 
> > Real deprecation is coming with 3.13 as `@deprecated` decorator. You could use typing_extension.deprecated right now but yes you either need to wrap it manually in pyx or add a new wrap-directive for autowrap (since cdef functions in pxds cannot be decorated arbitrarily.
> > 
> > Potentially the sphinx directive in the docstrings is enough for good IDEs
> 
> Does something multiline work?
> ```
>             fromString(String seq) # wrap-attach:AASequence 
>             # wrap-doc:
>               #  :param seq: Construct AASequence from String (e.g., "PEPTIDER")
>               #  .. deprecated:: 3.2
>               #     Use :py:func:`AASequence(String)` instead.
>               #  
> ```

I think so. There are some subtleties but it will work somehow.

**Comment by timosachsenberg** [comment]
> You could actually do it right now. When autowrap generates the pyi, you can check the docstring. If it contains `.. deprecated:` you add `@typing_extension.deprecated` before the type annotation.

you mean adding it to autowrap code generation?

**Comment by jpfeuffer** [comment]
Yes, e.g. here for enums:
https://github.com/OpenMS/autowrap/blob/master/autowrap%2FCodeGenerator.py#L553

```
if doc.contains(".. deprecated:"):
   stub_code.add("@typing_extensions.deprecated")
```

**Comment by timosachsenberg** [comment]
added an issue https://github.com/OpenMS/autowrap/issues/187

**Comment by timosachsenberg** [comment]
Completes this PR for now. I think we can do an official deprecation once we have it supported in autowrap (not sure though if using the sphinx syntax is the right way to go).

**Pull request merged**